lib/core-integration/src/Test/Integration/Framework/DSL.hs

```
@@ -643,6 +658,39 @@ direction =
     _set :: HasType (ApiT Direction) s => (s, Direction) -> s
     _set (s, v) = set typed (ApiT v) s
 
+insertedAt :: HasType (Maybe ApiBlockData) s => Lens' s (Maybe Text)
+insertedAt =
+    lens _get _set
+  where
+    utcToText = utcTimeToText iso8601ExtendedUtc
+    _get :: HasType (Maybe ApiBlockData) s => s -> (Maybe Text)
+    _get mb = case (view typed mb) of
+        Just b -> Just (utcToText (time b))
+        Nothing -> Nothing
+    _set :: HasType (Maybe ApiBlockData) s => (s, (Maybe Text)) -> s
+    _set (s, v) = case v of
+        Just t -> Just (set typed (ApiBlockData (textToUTCTime t) mempty) s)
+        Nothing -> Nothing
```

Hi @piotr-iohk!

Given that your other lenses have non-textual types, I wonder why we couldn't instead define `insertedAt` with the following type:

```hs
insertedAt :: HasType (Maybe ApiBlockData) s => Lens' s (Maybe UTCTime)
```

That way, we wouldn't need to have the silent default in the `textToUTCTime` function. The reason I worry about silent defaults is that they could mask errors introduced by people modifying the code in future.

However, if using the `Text` type is more useful, then perhaps we could have `_set` use a default of `Nothing` if the given text is invalid, which is compatible with the original type you've used (`Maybe Text`).

What do you think?

---

I was going initially for `Text` as I was thinking it'd be easier to assert in tests, but maybe `UTCTime` will be better, we'll see when I get to the tests part... First I need to make this freaking lens work ;)

> Given that your other lenses have non-textual types,
 
btw. actually there is already a date-time lens that is a `Text` type -> `passphraseLastUpdate`

>a default of Nothing if the given text is invalid,

If I go for `Text` I'll try this.

---

OK, went for:

```haskell
insertedAt :: HasType (Maybe ApiBlockData) s => Lens' s (Maybe UTCTime)
```